src/funnel.lodash-debounce-with-cached-value.test.ts

```diff
@@ -0,0 +1,252 @@
+ * Remeda `funnel` function. While migrating from Lodash you can copy this
+ * function as-is into your codebase and use it as a drop-in replacement; but
+ * we recommend eventually inlining the call to `funnel` so you can adjust the
+ * function to your specific needs.
+ *
+ * This is a more complex implementation which respects Lodash capability to
+ * track the return value of the callback function. In most cases you are more
+ * likely to prefer the simpler reference implementation `debounce` which can
+ * be found in the other test file.
+ *
+ * The following tests in this file are based on the Lodash tests for debounce.
+ * They have been adapted to work with our testing framework, have been fixed
+ * or expanded slightly were it felt necessary, and have been modernized for
+ * better readability. The names of the test cases have been preserved to ease
+ * comparing them to the original tests. Tests that are unrelated to the cache
+ * capability have been removed to avoid duplication with the other test file.
+ *
+ * Note that this means that whenever Lodash offered a concrete spec, we made
+ * sure our reference implementation respects it, but there might be untested
+ * use-cases that would have differing runtime behaviors.
+ *
+ * @see Lodash Documentation: https://lodash.com/docs/4.17.15#debounce
+ * @see Lodash Implementation: https://github.com/lodash/lodash/blob/4.17.21/lodash.js#L10372
+ * @see Lodash Tests: https://github.com/lodash/lodash/blob/4.17.21/test/test.js#L4187
+ * @see Lodash Typing: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/lodash/common/function.d.ts#L374
+ */
+function debounceWithCachedValue<F extends (...args: any) => any>(
+  func: F,
+  wait = 0,
+  {
+    leading = false,
+    trailing = true,
+    maxWait,
+  }: {
+    readonly leading?: boolean;
+    readonly trailing?: boolean;
+    readonly maxWait?: number;
+  } = {},
+) {
+  let cachedValue: ReturnType<F> | undefined;
+
+  const { call, flush, cancel } = funnel(
+    (args: Parameters<F>) => {
+      if (!leading && !trailing) {
+        // In Lodash you can disable both the trailing and leading edges of the
+        // debounce window, effectively causing the function to never be
+        // invoked. Remeda uses the invokedAt enum exactly to prevent such a
+        // situation; so to simulate Lodash we need to only pass the callback
+        // when at least one of them is enabled.
+        return;
+      }
+
+      // Funnel provides more control over the args, but lodash simply passes
+      // them through, to replicate this behavior we need to spread the args
+      // array maintained via the reducer below.
+      // Also, every time the function is invoked the cached value is updated.
+      cachedValue = func(...args) as ReturnType<F>;
+    },
+    {
+      // Debounce stores the latest args it was called with for the next
+      // invocation of the callback.
+      reducer: (_, ...args: Parameters<F>) => args,
```

Comment:
**question:** does this work when the callback is called with `undefined`?

Reply:
args is an array, if a function is *called* with an arg of `undefined` this would be `[undefined]` which is a non-trivial value and should work.